helios-client/src/main/java/com/spotify/helios/client/AuthenticatingHttpConnector.java

```diff
@@ -0,0 +1,178 @@
+
+import com.spotify.helios.client.tls.SshAgentSSLSocketFactory;
+import com.spotify.helios.common.HeliosException;
+import com.spotify.sshagentproxy.AgentProxy;
+import com.spotify.sshagentproxy.Identity;
+
+import org.apache.http.conn.ssl.DefaultHostnameVerifier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.ConnectException;
+import java.net.HttpURLConnection;
+import java.net.SocketTimeoutException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.UnknownHostException;
+import java.util.Deque;
+import java.util.List;
+import java.util.Map;
+
+import javax.net.ssl.HostnameVerifier;
+import javax.net.ssl.HttpsURLConnection;
+import javax.net.ssl.SSLSession;
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static java.net.HttpURLConnection.HTTP_FORBIDDEN;
+import static java.net.HttpURLConnection.HTTP_UNAUTHORIZED;
+
+/**
+ * HttpConnector that wraps another connector to add the necessary hooks into HttpURLConnection for
+ * our SSH/TLS-based authentication.
+ */
+public class AuthenticatingHttpConnector implements HttpConnector {
+
+  private static final Logger log = LoggerFactory.getLogger(AuthenticatingHttpConnector.class);
+
+  // TODO (mbrown): username is here to pass to SshAgentSSLSocketFactory, could be handled nicer
+  private final String user;
+  private final Optional<AgentProxy> agentProxy;
+  private final List<Identity> identities;
+  private final EndpointIterator endpointIterator;
+
+  private final HttpConnector delegate;
+
+  public AuthenticatingHttpConnector(final String user,
+                                     final Optional<AgentProxy> agentProxy,
+                                     final List<Identity> identities,
+                                     final EndpointIterator endpointIterator,
+                                     final HttpConnector delegate) {
+    this.user = user;
+    this.agentProxy = agentProxy;
+    this.identities = identities;
+    this.endpointIterator = endpointIterator;
+    this.delegate = delegate;
+  }
+
+  @Override
+  public HttpURLConnection connect(final URI uri, final String method, final byte[] entity,
+                                   final Map<String, List<String>> headers) throws HeliosException {
+    final Endpoint endpoint = endpointIterator.next();
+    final String endpointHost = endpoint.getUri().getHost();
+
+    // convert the URI whose hostname portion is a domain name into a URI where the host is an IP
+    // as we expect there to be several different IP addresses besides a common domain name
+    final URI ipUri;
+    try {
+      ipUri = toIpUri(endpoint, uri);
+    } catch (URISyntaxException e) {
+      throw new HeliosException(e);
+    }
+
+    final Deque<Identity> ids;
+    if (ipUri.getScheme().equalsIgnoreCase("https")) {
+      ids = Queues.newArrayDeque(identities);
+    } else {
+      ids = Queues.newArrayDeque();
```

Comment:
bug 1: this used to be `Queues.newArrayDeque(null)` which throws a NullPointerException as Guava tries to add all of a null `Iterable` to the queue